src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -275,6 +277,15 @@ private static ISymbol GetSymbolsToPullUp(MemberAnalysisResult analysisResult)
                         solution.GetDocumentId(syntax.SyntaxTree),
                         cancellationToken).ConfigureAwait(false);
 
+                    // add import symbol annotations for all the code we may be moving
+                    newDestination = newDestination.WithAdditionalAnnotations(
+                        syntax.DescendantNodesAndSelf()
+                        .Select((node) => semanticModel.GetTypeInfo(node, cancellationToken).Type)
+                        .Where((type) => type != null &&
+                                !type.IsErrorType())
+                        .Distinct()
+                        .Select((type) => SymbolAnnotation.Create(type)));
```

Comment:
I think it's worse than this.  First, we want to check any start of expression (which could be identifiers, generic names, etc.).  Second, things like extension methods are brought in by using statements, so we need to bind methods.

AN alternate approach that sidesteps this (though introduces its own problems is), is to avoid this kind of binding entirely.

Instead:

1. pull *all* the usings up to the destination location that are not already in the destination.
2. mark all existing usings in teh destination location somehow (or, alternatively, mark only the usings that were pulled up.
3. remove all unused usings as the final step *but* filter that only to the set that was added.

Effectively, this will leave us with only the set that was needed to bind the code that was moved.  THe downside is though that some of the added usings may cause conflicts in the destination due to ambiguous references.  However, that was already possible with the approach we're talking here, so maybe it's not too bad.

Reply:
: ) Just one point added, It is ok if you finally need to convert to non-static language service. I remember when this class is originally written, we have a discussion around whether it should be a static helper or abstract class -> C# impl/VB impl.

Reply:
Also, a document is passed in. So you can always acquire language services from that.

Reply:
One other option to consider would be to use the simplifier to expand (and later collapse) the moved nodes into using their fully qualified names. This might get around the destination conflicts problem as well, although that may need more testing.